File: android/src/com/android/tools/idea/gradle/project/model/ide/android/IdeAndroidArtifact.java
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.gradle.project.model.ide.android;

import com.android.annotations.Nullable;
import com.android.builder.model.*;
import com.android.ide.common.repository.GradleVersion;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.*;

/**
 * Creates a deep copy of {@link AndroidArtifact}.
 */
public class IdeAndroidArtifact extends IdeBaseArtifact implements AndroidArtifact {
  @NotNull private final Collection<AndroidArtifactOutput> myOutputs;
  @NotNull private final String myApplicationId;
  @NotNull private final String mySourceGenTaskName;
  @NotNull private final Collection<File> myGeneratedResourceFolders = new ArrayList<>();
  @NotNull private final Map<String, ClassField> myBuildConfigFields = new HashMap<>();
  @NotNull private final Map<String, ClassField> myResValues = new HashMap<>();
  @NotNull private final InstantRun myInstantRun;
  @Nullable private final String mySigningConfigName;
  @Nullable private final Set<String> myAbiFilters;
  @Nullable private final Collection<NativeLibrary> myNativeLibraries;
  private final boolean mySigned;

  public IdeAndroidArtifact(@NotNull AndroidArtifact artifact, @NotNull ModelCache modelCache, @NotNull GradleVersion gradleVersion) {
    super(artifact, modelCache, gradleVersion);

    myOutputs = copy(artifact.getOutputs(), modelCache, IdeAndroidArtifactOutput::new);

    myApplicationId = artifact.getApplicationId();
    mySourceGenTaskName = artifact.getSourceGenTaskName();
    myGeneratedResourceFolders.addAll(artifact.getGeneratedResourceFolders());

    Map<String, ClassField> buildConfigFields = artifact.getBuildConfigFields();
    buildConfigFields.forEach((name, classField) -> myBuildConfigFields.put(name, new IdeClassField(classField)));

    Map<String, ClassField> resValues = artifact.getResValues();
    resValues.forEach((name, classField) -> myResValues.put(name, new IdeClassField(classField)));

    myInstantRun = new IdeInstantRun(artifact.getInstantRun());
    mySigningConfigName = artifact.getSigningConfigName();

    Set<String> abiFilters = artifact.getAbiFilters();
    myAbiFilters = abiFilters != null ? new HashSet<>(abiFilters) : null;

    Collection<NativeLibrary> nativeLibraries = artifact.getNativeLibraries();
    if (nativeLibraries != null) {
      myNativeLibraries = copy(nativeLibraries, modelCache, IdeNativeLibrary::new);
    }
    else {
      myNativeLibraries = null;
    }

    mySigned = artifact.isSigned();
  }

  @Override
  @NotNull
  public Collection<AndroidArtifactOutput> getOutputs() {
    return myOutputs;
  }

  @Override
  @NotNull
  public String getApplicationId() {
    return myApplicationId;
  }

  @Override
  @NotNull
  public String getSourceGenTaskName() {
    return mySourceGenTaskName;
  }

  @Override
  @NotNull
  public Collection<File> getGeneratedResourceFolders() {
    return myGeneratedResourceFolders;
  }

  @Override
  @NotNull
  public Map<String, ClassField> getBuildConfigFields() {
    return myBuildConfigFields;
  }

  @Override
  @NotNull
  public Map<String, ClassField> getResValues() {
    return myResValues;
  }

  @Override
  @NotNull
  public InstantRun getInstantRun() {
    return myInstantRun;
  }

  @Override
  @Nullable
  public String getSigningConfigName() {
    return mySigningConfigName;
  }

  @Override
  @Nullable
  public Set<String> getAbiFilters() {
    return myAbiFilters;
  }

  @Override
  @Nullable
  public Collection<NativeLibrary> getNativeLibraries() {
    return myNativeLibraries;
  }

  @Override
  public boolean isSigned() {
    return mySigned;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    IdeAndroidArtifact artifact = (IdeAndroidArtifact)o;
    return mySigned == artifact.mySigned &&
           Objects.equals(myOutputs, artifact.myOutputs) &&
           Objects.equals(myApplicationId, artifact.myApplicationId) &&
           Objects.equals(mySourceGenTaskName, artifact.mySourceGenTaskName) &&
           Objects.equals(myGeneratedResourceFolders, artifact.myGeneratedResourceFolders) &&
           Objects.equals(myBuildConfigFields, artifact.myBuildConfigFields) &&
           Objects.equals(myResValues, artifact.myResValues) &&
           Objects.equals(myInstantRun, artifact.myInstantRun) &&
           Objects.equals(mySigningConfigName, artifact.mySigningConfigName) &&
           Objects.equals(myAbiFilters, artifact.myAbiFilters) &&
           Objects.equals(myNativeLibraries, artifact.myNativeLibraries);
  }

  @Override
  public int hashCode() {
    return Objects.hash(myOutputs, myApplicationId, mySourceGenTaskName, myGeneratedResourceFolders, myBuildConfigFields, myResValues,
                        myInstantRun, mySigningConfigName, myAbiFilters, myNativeLibraries, mySigned);
  }

  @Override
  public String toString() {
    return "IdeAndroidArtifact{" +
           "myOutputs=" + myOutputs +
           ", myApplicationId='" + myApplicationId + '\'' +
           ", mySourceGenTaskName='" + mySourceGenTaskName + '\'' +
           ", myGeneratedResourceFolders=" + myGeneratedResourceFolders +
           ", myBuildConfigFields=" + myBuildConfigFields +
           ", myResValues=" + myResValues +
           ", myInstantRun=" + myInstantRun +
           ", mySigningConfigName='" + mySigningConfigName + '\'' +
           ", myAbiFilters=" + myAbiFilters +
           ", myNativeLibraries=" + myNativeLibraries +
           ", mySigned=" + mySigned +
           '}';
  }
}

File: project-system-gradle-models/src/com/android/tools/idea/gradle/model/IdeAndroidArtifact.kt
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.gradle.model

import java.io.File
import java.io.Serializable

interface IdeAndroidArtifactCore : Serializable, IdeBaseArtifactCore {

  /**
   * The application ID of this artifact.
   *
   * Known for:
   *  - Application plugin main artifacts
   *  - AndroidTest components of all project types
   *  - Test-only plugin main artifacts
   *
   *  Not included (null) for:
   *   - Library plugin main artifacts, as no APK is produced
   *   - UnitTest components, also as no APK is produced
   *   - Dynamic feature plugin main artifacts, as the application ID comes from the base
   *     application, and is therefore not available in dynamic feature projects during
   *     configuration. In this case Android Studio must look at the dependency graph to find the
   *     base application to find this value.
   */
  val applicationId: String?

  /**
   * Returns the name of the [IdeSigningConfig] used for the signing. If none are setup or if
   * this is not the main artifact of an application project, then this is null.
   */
  val signingConfigName: String?

  /**
   * Returns whether the output file is signed. This can only be true for the main apk of an
   * application project.
   */
  val isSigned: Boolean

  val generatedResourceFolders: Collection<File>

  /**
   * Map of generated res values where the key is the res name. This method is deprecated and will
   * always return an empty map
   */
  @Deprecated("Returns empty map")
  val resValues: Map<String, IdeClassField>

  /**
   * Returns a list of additional APKs that need to installed on the device for this artifact to
   * work correctly.
   *
   *
   * For test artifacts, these will be "buddy APKs" from the `androidTestUtil`
   * configuration.
   */
  val additionalRuntimeApks: List<File>

  /**
   * Returns the test options only if the variant type is testing.
   */
  val testOptions: IdeTestOptions?

  val abiFilters: Set<String>

  val buildInformation: IdeBuildTasksAndOutputInformation

  /**
   * Returns the code shrinker used by this artifact or null if no shrinker is used to build this
   * artifact.
   */
  val codeShrinker: CodeShrinker?

  /**
   * Details about privacy sandbox SDK consumption.
   *
   * Null if the feature is not enabled, but may be present even if there are
   * no privacy sandbox SDKs that need to be deployed.
   */
  val privacySandboxSdkInfo: IdePrivacySandboxSdkInfo?

  /**
   * Returns all the [IdeModelSyncFile] for this artifact.
   *
   * @return a list of [IdeModelSyncFile]
   * @since 7.2
   */
  val modelSyncFiles: Collection<IdeModelSyncFile>
  val compileClasspathCore: IdeDependenciesCore
  val runtimeClasspathCore: IdeDependenciesCore

  /**
   * Files listing any D8 backported desugared methods or core library desugared methods for this artifact
   */
  val desugaredMethodsFiles: Collection<File>
}

interface IdeAndroidArtifact : IdeAndroidArtifactCore, IdeBaseArtifact